tools/indexscan/main.go

```diff
@@ -251,6 +252,103 @@ func (s CommitsNearPublishStrategy) Search(ctx context.Context, r *git.Repositor
 	return
 }
 
+// ExhaustiveTreeSearchStrategy
+type ExhaustiveTreeSearchStrategy struct {
+}
+
+// Search returns the set of matching commits along with the number of matches.
+func (s ExhaustiveTreeSearchStrategy) Search(ctx context.Context, r *git.Repository, hashes []string) (closest []string, matched, total int, err error) {
+	// Find subset of file hashes that are present in the repo.
+	files := make(map[plumbing.Hash]int)
+	var i int
+	for _, h := range hashes {
+		_, err = r.BlobObject(plumbing.NewHash(h))
+		if err == plumbing.ErrObjectNotFound {
+			// Skip files not present in repo.
+			err = nil
+		} else if err != nil {
+			return
+		} else {
+			files[plumbing.NewHash(h)] = i
+			i++
+		}
+	}
+	// Iterate over repo trees to create an ordering.
+	trees := make(map[plumbing.Hash]int)
+	i = 0
+	for ti, _ := r.TreeObjects(); ; {
+		t, err := ti.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			panic(err)
```

Comment:
If the above loop just returns err instead of panic, should this loop do the same?

Reply:
I think the way it's assigned (`t, err := ti.Next()`) means the function's err value won't be updated and the name will be shadowed instead🤷

Reply:
Ah, gotcha.